### PR TITLE
ATOMIC-TS-WEBVIEWEDITOR

### PR DESCRIPTION
**Pull request by shaddockh**
This converts the editor code that lives inside the WebView to typescript to set the stage for further work. 

 In order to do this, Typescript needed to be upgraded to version 1.8 in order to bundle the code into a single file.  There is also a dependency on https://github.com/systemjs/systemjs which is a module loader for the web that is supported by the typescript configuration option: ```module: "system"```

The code has a single entrypoint called interop.ts which sets up the functions on ```window``` that the host needs to be able to reference.  Any api calls the client can call or the host can call should be placed into that interop file and wired up to the ```window``` object as necessary.

Additionally, since CEF now supports the ES6 spec, the code that is being emitted is emitted as ES6 javascript.  This allows for native support of some of the es6 features such as fat arrows, promises, etc.

Finally, TS 1.8 detected some errors in the editor that were not caught in 1.7x, so those were fixed.

### CODE REVIEW

File: Data/AtomicEditor/CodeEditor/Editor.html [hunk 1]
```diff
@@ -25,55 +25,10 @@
...
-    xmlHttp.onreadystatechange = function() {
-      if (xmlHttp.readyState == 4 && xmlHttp.status == 200)
-      codeLoaded(xmlHttp.responseText);
-    }
-    xmlHttp.open("GET", codeUrl, true); // true for asynchronous
-    xmlHttp.send(null);
-  }
-
+    System.import('./interop');
   </script>
```
Comment (JoshEngebretson) [inline Data/AtomicEditor/CodeEditor/Editor.html]: Many lines replaced with one, neat! :)

File: Script/AtomicEditor/editor/Preferences.ts [hunk 2]
```diff
@@ -32,7 +32,7 @@ class Preferences {
 
     updateRecentProjects(write: boolean = false): void {
 
-        for (var i in this._prefs.recentProjects) {
+        for (var i = 0; i < this._prefs.recentProjects.length; i++) {
             var path = this._prefs.recentProjects[i];
```
Comment (JoshEngebretson) [inline Script/AtomicEditor/editor/Preferences.ts]: Whoa, what's the story here?

File: Script/AtomicWebViewEditor/interop.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        };
+        xmlHttp.open("GET", codeUrl, true); // true for asynchronous
+        xmlHttp.send(null);
+    }).then(function(src: string) {
+        codeLoaded(src, fileExt);
+    });
+}
+
+// Set up the window object so the host can call into it
+window.loadCode = loadCode;
```
Comment (JoshEngebretson) [inline Script/AtomicWebViewEditor/interop.ts]: TS is better than JS any day of the week

File: Script/AtomicWebViewEditor/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    ],
+    "atom": {
+        "rewriteTsconfig": true
+    },
+    "files": [
+        "./editor/editor.ts",
+        "./editor/editorConfig.ts",
+        "./interop.ts",
+        "./typings/ace.d.ts",
+        "./typings/AtomicQuery.d.ts"
```
Comment (JoshEngebretson) [inline Script/AtomicWebViewEditor/tsconfig.json]: oooooh, AtomicQuery.d.ts!

File: Build/Scripts/BuildCommon.js [hunk 5]
```diff
@@ -80,6 +80,7 @@ namespace('build', function() {
         if (node) {
           // compile
           cmds.push(node + " " + tsc + " -p ./Script");
+          cmds.push(node + " " + tsc + " -p ./Script/AtomicWebViewEditor");
```
Comment (JoshEngebretson) [inline Build/Scripts/BuildCommon.js]: wow, more compiling, this is so cool :+1:

File: Build/Scripts/BuildCommon.js [hunk 6]
```diff
@@ -94,7 +95,8 @@ namespace('build', function() {
             });
           });
 
-          lintTask.invoke("./Script/AtomicEditor/**/*.ts", false);
+          lintTask.invoke("{./Script/AtomicEditor/**/*.ts,./Script/AtomicWebViewEditor/**/*.ts}", false);
+          //lintTask.invoke("./Script/AtomicWebViewEditor/**/*.ts", false);
```
Comment (JoshEngebretson) [inline Build/Scripts/BuildCommon.js]: Dead code?

File: Script/AtomicEditor/editor/Preferences.ts (hunk 2 above)
Comment (shaddockh) [inline Script/AtomicEditor/editor/Preferences.ts]: this._prefs.recentProjects is a ```string[]```  not an object, so you will want to iterate over it like any other array, and not like a key/value object.  TS 1.8 flagged this as an error.

File: Build/Scripts/BuildCommon.js (hunk 6 above)
Comment (shaddockh) [inline Build/Scripts/BuildCommon.js]: yep.. fixing that.

**Comment by JoshEngebretson** [comment]
Windows decided to reboot mid-build...

@AtomicBuildBot retest this please

**Comment by JoshEngebretson** [comment]
Great stuff, landing and will address the CI stuff in a followup PR shortly! :airplane:

**Comment by JoshEngebretson** [comment]
Easy peasy: https://github.com/AtomicGameEngine/AtomicGameEngine/pull/642/files

:1:

**Comment by shaddockh** [comment]
Awesome, so is the editor only built on OSX in CI?

**Pull request merged**